site/source/_posts/2017/11/17-Dojo-2-Beta-4/Dojo-2-Beta-4.md

```diff
@@ -0,0 +1,263 @@
+
+## Stores re-architecture
+
+The current API in Dojo 1 for managing state focused on a CRUD type API.  Originally in Dojo 2 we directly adapted that API as part of _stores_.  Recently though, the concept of state containers has taken front and centre and we knew we wanted to create a solution that aligned to that architecture.
+
+Instead of CRUD type API, it focuses on three main concepts:
+
+* _Operation_ which is a granular state mutation based on the JSON Patch API
+* _Command_ a semantically aligned function which returns a set of operations to apply
+* _Process_ a function that groups commands into a _process flow_ to represent a complete behaviour
+
+You can read more about the new API at [`dojo/stores`](https://github.com/dojo/stores).
+
+## A new approach to shims and polyfills
+
+We have rationalised the way we introduce compatibility across environments as well as improvements to the build system to optimize these when building an application.  We realised that some packages were wanting to depend upon 3rd party polyfills but we did not have a consistent way to introduce them as well as elide them if not needed.
+
+Now `@dojo/shim/main` loads all the polyfills needed to provide the expected base functionality and `@dojo/shim/browser` will load polyfills for a browser environment.  Most end developers will not need to worry about this as the other Dojo 2 packages will ensure that these modules get loaded.
+
+Of course if you are building an application that is targetting only certain platforms, these polyfills will simply result in a non-operation, but will still be part of the bundle.  We have added an additional static build optimisation feature, where a specially formed pragma will denote the feature that the next import will be polyfilling, and if that feature is not required based on the build targets, it will be removed from the build.  More information can be found in the [documentation for the feature](https://github.com/dojo/webpack-contrib#elided-imports).
+
+## Iterators
+
+Previously we had provided a functional replacement for `for...of` constructs so that developers could author code that worked in older browsers as well as we provided shims for `Symbol` and the built in ES6 iterators.  In TypeScript 2.3 though, a down level emit for iterators was introduced.  We have restructured our packages to use `for...of` instead of the `@dojo/shim/iterator.forOf()` function and will likely remove `forOf()` in the future.  We recommend users of Dojo 2 also migrate to this syntax.
+
+## TypeScript helpers
+
+When emitting to JavaScript from TypeScript, all of our run-time modules included any of the TypeScript helpers.  This causes the bundle size to increase because these helpers would be written out multiple times.  In Beta 4, we have changed to import our helpers from `tslib` and automatically include these helpers once as part of the CLI build tool.
```

Comment:
When emitting to JavaScript from TypeScript, all of our run-time modules included any of the TypeScript helpers.  This causes the bundle size to increase because these helpers would be written out multiple times.  In Beta 4, we have changed to import our helpers from `tslib` and automatically include these helpers once as part of the CLI build tool. -> When emitting to JavaScript from TypeScript, each run-time module included all TypeScript helpers needed by that module.  This increases the bundle size because these helpers would be duplicated many times.  In Beta 4, we import helpers from `tslib` and automatically include these helpers once as part of the CLI build tool, eliminating any duplication of TS helper code.